sailfishusermanager.cpp

```diff
@@ -348,3 +349,65 @@ bool SailfishUserManager::hasAccessRights(uint uid_to_modify) {
 
     return true;
 }
+
+void SailfishUserManager::userSwitch(uint uid)
+{
+    if (!hasAccessRights(SAILFISH_UNDEFINED_UID)) {
+        auto message = QStringLiteral("Access denied");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::AccessDenied, message);
+        return;
+    }
+
+    uid_t currentUid;
+    if (sd_seat_get_active("seat0", nullptr, &currentUid) < 0) {
+        auto message = QStringLiteral("Failed to get current user id");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    QDBusInterface systemd(QStringLiteral("org.freedesktop.systemd1"), QStringLiteral("/org/freedesktop/systemd1"),
+                           QStringLiteral("org.freedesktop.systemd1.Manager"), QDBusConnection::systemBus());
+
+    // Stop user service
+    QDBusReply<QDBusObjectPath> reply = systemd.call(QStringLiteral("StopUnit"),
+                                                     QStringLiteral("user@%1.service").arg(currentUid),
+                                                     QStringLiteral("replace"));
+    if (!reply.isValid()) {
+        auto message = QStringLiteral("Failed to stop current user session");
+        qCWarning(lcSUM) << message;
+        sendErrorReply(QDBusError::Failed, message);
+        return;
+    }
+
+    // Stop autologin service
+    reply = systemd.call(QStringLiteral("StopUnit"), QStringLiteral("autologin@%1.service").arg(currentUid),
+                         QStringLiteral("replace"));
+    if (!reply.isValid()) {
```

Comment:
This service is not serving D-Bus requests why this blocking call is pending, right? Could this cause any problems? Timeout is an obvious one, but could there be something trickier like a deadlock (A calls B, B calls C, C calls A - if all calls are blocking, it's a deadlock)?

Reply:
There's a default timeout for call, 25 secs according to docs.

Reply:
It does sound to me that it's a bit bad idea to make these synchronous although it probably works alright in the end most of the time now. I'm not 100% sure that there won't be something calling this daemon while it is ending the session. Currently there isn't but I think that could change in later stages.

Reply:
Switched to async calls